### ISSUE TITLE
$layout unexpectedly initialises twice in development

### ISSUE THREAD

**Issue by intrikate** [sveltejs/kit#1214]
**Describe the bug**
Updating `$layout.svelte` and reloading the page makes it initialise and mount twice: once on reload and once more on next navigation. This will repeat on every page reload until the dev server is restarted.

**To Reproduce**
- Clone and run: https://github.com/intrikate/sveltekit-layout-init-issue.
- Navigate to `localhost:3000`, open the console.
- Update `src/routes/$layout.svelte` in any way (e.g. delete the empty comment in line 4).
- Save and **reload** the page. (HMR on its own does not trigger this behaviour.)
- Use the links for navigation, count the console statements about the layout lifecycle.

**Expected behavior**
The layout component should mount once and persist between routes, with no unmounts and reinitialisation.

**Information about your SvelteKit Installation:**

<details>
  <summary>Diagnostics</summary>

  System:
    OS: macOS 11.2.3
    CPU: (4) x64 Intel(R) Core(TM) i5-4288U CPU @ 2.60GHz
    Memory: 55.59 MB / 8.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 14.16.0 - /usr/local/bin/node
    Yarn: 1.22.10 - /usr/local/bin/yarn
    npm: 6.14.11 - /usr/local/bin/npm
  Browsers:
    Chrome: 90.0.4430.85
    Firefox Developer Edition: 89.0
    Safari: 14.0.3
  npmPackages:
    @sveltejs/kit: next => 1.0.0-next.89 
    svelte: ^3.37.0 => 3.37.0 
    vite: ^2.2.2 => 2.2.2 
</details>

**Severity**
High, as it leads to obscure lifecycle-related bugs.

**Additional Context**
Complex transitions and per-route layout tweaks are difficult or outright impossible without the layout to coordinate them.

**Comment by UltraCakeBakery** [comment]
I've been experiencing this behavior since `.83`. It looks like a route navigation is required for hydration to actually effect the dom. As if the hydration the first time doesn't complete itself right before it is supposed to update the DOM and doesn't remember that it already did _something_.

**Comment by UltraCakeBakery** [comment]
I also now have my components inside my `$layout` being added to the DOM twice. Once inside my layout where they are supposed to mount (as specified in my code) and outside my hydration target.

**Comment by intrikate** [comment]
> I've been experiencing this behavior since `.83.`

I can still reproduce it in .82. Apparently, the most recent one where I can’t is .71.

I was actually thinking of commenting on #1171, but opted to reduce it further and specify triggering conditions (wasn’t sure those matched between the two). Overall, #1171 is pretty similar to the real case where it first bit me. Add stores to the mix (as, say, in the [proposed solution](https://github.com/sveltejs/kit/issues/1110#issuecomment-822345891) to your #1110), throw in store-triggered transitions, and you’re a queen bee.

**Comment by chuckrector** [comment]
I've been porting a Sapper app over to SvelteKit and have been experiencing something similar. It may be a separate issue but it does share the duplication aspect. Here is a minimal repro: https://github.com/chuckrector/kit-dupes

I had copy/pasted a large Sapper component over into my SvelteKit project and noticed that only a certain part of it was getting duplicated on reload. I whittled the entire file down to this:
```html
SvelteKit <a href="/"><a href="/">test</a></a>
```
The repro linked above was created by following the instructions on the landing page, as a fresh SvelteKit skeleton project with no other frills and replaces `index.svelte` with the above content. There is no `$layout.svelte`. It seems to happen in both development mode and in preview mode.

**Comment by intrikate** [comment]
@chuckrector How long have you been having this? Nested anchors are tricky, as would be any invalid HTML, but otherwise it looks pretty much like #1288.

**Comment by chuckrector** [comment]
@intrikate I started yesterday after Rich retweeted some of the new changes in 3.38. The issue you've linked appears to be precisely it. Thank you! Having read that it was resolved, I tried creating a new SvelteKit project with the latest changes and can no longer reproduce the issue. Li Hau is quick on his feet. 🏃‍♂️💨

**Comment by asv7c2** [comment]
@intrikate 
Looks like bug is fixed now.

**Comment by intrikate** [comment]
@asv7c2, are you’re referring to https://github.com/sveltejs/svelte/issues/6274 (formerly #1288)? This one isn’t actually related and still reproducible with .94 and Svelte 3.38.1.

**Comment by asv7c2** [comment]
@intrikate 
No, i checked your repo, layout is mounted only once.

**Comment by intrikate** [comment]
@asv7c2, now that’s strange, I can still see it. You did update the file, right?

**Comment by asv7c2** [comment]
@intrikate 
Sorry, bug is here. I just not editing $layout.

**Comment by Litarvan** [comment]
I'm having the same issue, this prevents my css transitions in the layout to happen between route changes (as the layout component is re-initialized).

**Comment by intrikate** [comment]
I now believe we’re looking at misplaced layout reset. For whatever reason, modifying `$layout.svelte` and reloading the page causes the layout passed to `start()` to fail the identity check on subsequent navigation. The following correctly evaluates to `false` for the hot-replaced layout component, but after the reload `module !== previous.module` becomes `true`, apparently prompting the renderer to destroy and recreate the layout.

https://github.com/sveltejs/kit/blob/bbbbe89f244caaf926daecaf5f248d89e2e3a9a3/packages/kit/src/runtime/client/renderer.js#L534-L541

The layout component is actually passed with hydration data, but somehow it doesn’t match the same layout imported by the renderer. I’m hoping maybe someone can explain this.

**Comment by wiesson** [comment]
I'm experiencing some similar issues and I'm wondering if they are related.

My flow:

1. /auth/signup/profile (`__layout.reset.svelte`), checks in onMount if user is not authenticated, `goto("/auth/login")`
2. /auth/login -> logs in, `goto("/auth/signup/profile")`
3. /auth/signup/profile - onMount is not called again

From your description, I just guess that goto() leaves the page too early so that hydration or some other processes can't finish. I also saw duplicated pages on top of each other (old on on top, new one bellow) - and as always happened only in prod, not in dev 😅

**Comment by wiesson** [comment]
Thanks @intrikate - I will try to reproduce it and create a separate issue
